docs/_docs/03-ug-mutable-dataset.md

```diff
@@ -87,7 +89,83 @@ hs.refreshIndex("empIndex", "incremental")
+if possible, and replacing them with fewer larger files that capture exact same index records. This process is similar to
+compaction in append-only log structured merge index structures.
+
+You should note that the `optimizeIndex` command is a best effort to modify index files layout and its
+final outcome depends on how index records are stored in existing index files.
+Two or more index files can be merged with each other, if and only if they all have index records which belong to the
+same bucket (according to index configuration).
+When running `optimizeIndex` command, Hyperspace tries to find such groups of index files and merge them together.
+However, if for each given bucket, all index records are already stored in a single index file (i.e. we dont have several index
+files for a bucket), then final physical layout of index files after running optimize would be similar to the original
+layout. An example of such an index is an index right after creation or full refresh.    
+
+Currently, there are two optimize modes available for an index: `"quick"` and `"full"`. These modes differ with each other
+in terms of the subset of index files they identify and try to merge.
+ 
+#### Optimize Index - Quick Mode
+Using `optimizeIndex` command with the `"quick"` mode on an index with many index files, due to incremental index refresh,
+causes Hyperspace look for index files which are smaller than a configurable size threshold and try to merge them.
+This mode tries to achieve a moderate query performance improvement through a fast optimize index process.
+The size threshold for an index file to be eligible for merging during quick optimization can be changed.
+Check the [configuration](02-ug-configuration.md) page to see how this threshold can be adjusted.
+Quick mode is the default mode for `optimizeIndex`.
+
+Assume you have an index with the name `empIndex`. After adding some data files to the dataset `empIndex` is created on
+and refreshing it in the incremental mode, you can optimize `empIndex` in the quick mode as below:
+
+Scala:
+```scala
+import com.microsoft.hyperspace._
+
+val hs = new Hyperspace(spark)
+hs.optimizeIndex("empIndex", "quick")
+``` 
+
+Python:
+
+```python
+from hyperspace import Hyperspace
+
+hs = Hyperspace(spark)
+hs.optimizeIndex("empIndex", "quick")
+```
+
+#### Optimize Index - Full Mode
+In the `"full"` mode, `optimizeIndex` command considers all existing index files as candidates for merging. Therefore,
+Hyperspace does a full scan on current index content and identifies groups of index files which have index records belonging
+to the same bucket according to the index configuration. It then replaces each group with a single index file created through
+merging all the files in that group together. This mode tries to achieve the best query performance improvement via a
+potentially slow optimize index process.
```

Comment:
Can you mention the difference between "full" optimize vs. "full" refresh? It is subtle, so it will be worthwhile to call it out.

Reply:
Added a paragraph on it.